airbyte-ci/connectors/metadata_service/lib/metadata_service/stale_metadata_report.py

```diff
@@ -0,0 +1,140 @@
+
+
+def _entry_should_be_on_gcs(metadata_dict: Mapping[str, Any]) -> bool:
+    if metadata_dict["supportLevel"] == "archived":
+        return False
+    if "-rc" in metadata_dict["dockerImageTag"]:
+        return False
+    return True
+
+
+def _get_latest_metadata_versions_on_github() -> Mapping[str, Any]:
+    github_token = os.getenv("GITHUB_TOKEN")
+    auth = Auth.Token(github_token)
+    github_client = Github(auth=auth)
+    repo = github_client.get_repo(REPOSITORY_NAME)
+
+    # Gets all files in the connectors folder, then filters for metadata.yaml files that are older than the grace period and appends the download url to a list.
+    contents = repo.get_contents("/airbyte-integrations/connectors/")
+    metadata_download_urls = []
+    while contents:
+        file_content = contents.pop(0)
+        if file_content.type == "dir":
+            contents.extend(repo.get_contents(file_content.path))
+        elif file_content.name == "metadata.yaml" and not _is_younger_than_grace_period(file_content):
+            metadata_download_urls.append(file_content.download_url)
+
+    metadata_dicts = []
+    for metadata_download_url in metadata_download_urls:
+        response = requests.get(metadata_download_url)
+        response.raise_for_status()
+        metadata_yaml = response.text
+        metadata_dict = yaml.safe_load(metadata_yaml)
+        metadata_dicts.append(metadata_dict)
+
+    latest_metadata_versions_on_github = {
+        metadata_dict["dockerRepository"]: metadata_dict["dockerImageTag"]
+        for metadata_dict in metadata_dicts
+        if _entry_should_be_on_gcs(metadata_dict)
+    }
+
+    return latest_metadata_versions_on_github
+
+
+def _get_latest_metadata_entries_on_gcs(bucket_name: str) -> Mapping[str, Any]:
+    storage_client = get_gcs_storage_client()
+    bucket = storage_client.bucket(bucket_name)
+
+    blobs: list[storage.Blob] = bucket.list_blobs(match_glob=f"{METADATA_FOLDER}/**/latest/{METADATA_FILE_NAME}")
+
+    latest_metadata_entries_on_gcs = {}
+    for blob in blobs:
+        metadata_dict = yaml.safe_load(blob.download_as_string().decode("utf-8"))
+        latest_metadata_entries_on_gcs[metadata_dict["dockerRepository"]] = metadata_dict["dockerImageTag"]
+
+    return latest_metadata_entries_on_gcs
+
+
+def _generate_stale_metadata_report(
+    latest_metadata_versions_on_github: Mapping[str, Any], latest_metadata_entries_on_gcs: Mapping[str, Any]
+) -> pd.DataFrame:
+    stale_connectors = []
+    for docker_repository, github_docker_image_tag in latest_metadata_versions_on_github.items():
+        gcs_docker_image_tag = latest_metadata_entries_on_gcs.get(docker_repository)
+        if gcs_docker_image_tag != github_docker_image_tag:
+            stale_connectors.append(
+                {"connector": docker_repository, "master_version": github_docker_image_tag, "gcs_version": gcs_docker_image_tag}
+            )
+
+    return pd.DataFrame(stale_connectors)
```

Comment:
curious: why do we need to use a dataframe for this? I.e. instead of just returning `stale_connectors` directly